### ISSUE TITLE
[Question] Why yank openvino 0.7.0/1?

### ISSUE THREAD

**Issue by oleflb** [intel/openvino-rs#133]
What is the reason 0.7.0 and 0.7.1 are yanked on crates.io? I could not find any information on that. This forces me to update the openvino dependency, which currently is hard for me since it doesnt seem to be possible anymore for tensors to be created from slices without copying the data.

**Comment by abrown** [comment]
I think you're looking for https://github.com/intel/openvino-rs/pull/125, in which @rahulchaphalkar removed the function you need after running across a crash. I tried to add some more detail to the commit message (https://github.com/intel/openvino-rs/commit/b6eacef9b5ab2cc964abbf02be3c74b0d7a948b1) but it boils down to "unsafety across the FFI." Looking at the download numbers on crates.io for those versions (very low still since they are quite new) and the potential for misuse and security vulnerabilities, I decided to just yank the crates. We published 0.7.2 without that function.

But you probably _need_ that function, so let's discuss a way to add something similar back in, but safely. We talked about potentially using Rust lifetimes and `PhantomData` to ensure that users keep the underlying tensor data alive as long as `Tensor`, but this caused extra lifetime complexity elsewhere. It still might be a valid approach, though. Are you interested in working on this?

**Comment by oleflb** [comment]
Yes, this function is what we used to avoid allocations :D. I think adding a lifetime to `Tensor` is perfectly valid, however users probably still want owned versions of `Tensor`. In `ndarray` they do something similar by having a generic `Container` attribute for `ArrayBase` which allows both for owned arrays and array views. Do you think something like this is suitable for the `Tensor` type here as well?

**Comment by abrown** [comment]
Yeah, that's a good idea too: kind of like [`Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html), we could have _owned_ and _borrowed_ variants. I think I like that better than the `PhantomData` idea because it's a bit more explicit about what's going on. We may still need to propagate lifetimes in a bunch of places; specifically, [`Core::read_model_from_buffer`](https://docs.rs/openvino/latest/openvino/struct.Core.html#method.read_model_from_buffer) accepts a `Tensor` and holds on to it for some indeterminate time (maybe until `Core::compile_model`, IIRC?).

**Comment by abrown** [comment]
@nbigaouette, the essence of the problem is that  [`Core::read_model_from_buffer`](https://docs.rs/openvino/latest/openvino/struct.Core.html#method.read_model_from_buffer) accepts a `Tensor` and holds on to it for some indeterminate time on OpenVINO's C side of the FFI but on the Rust side we will have dropped the slice. This is bad, since that memory could be reused for something else &mdash; memory corruption &mdash; which #125 prevents.

I think the best approach here is some way to represent both owned and borrowed tensor data; this could be either the explicit `Cow` approach I mentioned or a trait-based `ArrayView` approach like @oleflb suggested. Either approach seems viable, but a bit more discussion is needed and I haven't had time to get around to this. Are you interested in contributing here? We could discuss more on [Zulip](https://bytecodealliance.zulipchat.com).
